src/app/fdctl/run/tiles/fd_cswtch.c

```diff
@@ -0,0 +1,197 @@
+
+  FD_SCRATCH_ALLOC_INIT( l, scratch );
+  fd_cswtch_ctx_t * ctx = FD_SCRATCH_ALLOC_APPEND( l, alignof( fd_cswtch_ctx_t ), sizeof( fd_cswtch_ctx_t ) );
+
+  FD_TEST( topo->tile_cnt<FD_TOPO_MAX_TILES );
+
+  ctx->tile_cnt = topo->tile_cnt;
+  for( ulong i=0UL; i<topo->tile_cnt; i++ ) {
+    ulong * metrics = fd_metrics_join( fd_topo_obj_laddr( topo, topo->tiles[ i ].metrics_obj_id ) );
+
+    for(;;) {
+      ulong pid, tid;
+      if( FD_UNLIKELY( tile->id==i ) ) {
+        pid = fd_sandbox_getpid();
+        tid = fd_sandbox_gettid();
+      } else {
+        pid = fd_metrics_tile( metrics )[ FD_METRICS_GAUGE_TILE_PID_OFF ];
+        tid = fd_metrics_tile( metrics )[ FD_METRICS_GAUGE_TILE_TID_OFF ];
+        if( FD_UNLIKELY( !pid || !tid ) ) {
+          FD_SPIN_PAUSE();
+          continue;
+        }
+      }
+
+      char path[ 64 ];
+      FD_TEST( fd_cstr_printf_check( path, sizeof( path ), NULL, "/proc/%lu/task/%lu/status", pid, tid ) );
+      ctx->status_fds[ i ] = open( path, O_RDONLY );
+      ctx->metrics[ i ] = fd_metrics_tile( metrics );
+      if( FD_UNLIKELY( -1==ctx->status_fds[ i ] ) ) FD_LOG_ERR(( "open failed (%i-%s)", errno, strerror( errno ) ));
+      break;
+    }
+  }
+}
+
+static void
+unprivileged_init( fd_topo_t *      topo,
+                   fd_topo_tile_t * tile,
+                   void *           scratch ) {
+  (void)topo;
+
+  FD_SCRATCH_ALLOC_INIT( l, scratch );
+  fd_cswtch_ctx_t * ctx = FD_SCRATCH_ALLOC_APPEND( l, alignof( fd_cswtch_ctx_t ), sizeof( fd_cswtch_ctx_t ) );
+
+  ctx->next_report_nanos = fd_log_wallclock();
+
+  ulong scratch_top = FD_SCRATCH_ALLOC_FINI( l, 1UL );
+  if( FD_UNLIKELY( scratch_top > (ulong)scratch + scratch_footprint( tile ) ) )
+    FD_LOG_ERR(( "scratch overflow %lu %lu %lu", scratch_top - (ulong)scratch - scratch_footprint( tile ), scratch_top, (ulong)scratch + scratch_footprint( tile ) ));
+}
+
+static void
+before_credit( void *             _ctx,
+               fd_mux_context_t * mux ) {
+  (void)mux;
+
+  fd_cswtch_ctx_t * ctx = (fd_cswtch_ctx_t *)_ctx;
+
+  long now = fd_log_wallclock();
+  if( FD_UNLIKELY( now<ctx->next_report_nanos ) ) return;
+  ctx->next_report_nanos += REPORT_INTERVAL_MILLIS*1000L*1000L;
+
+  for( ulong i=0UL; i<ctx->tile_cnt; i++ ) {
+    if( FD_UNLIKELY( -1==lseek( ctx->status_fds[ i ], 0, SEEK_SET ) ) ) FD_LOG_ERR(( "lseek failed (%i-%s)", errno, strerror( errno ) ));
+
+    char contents[ 4096 ];
+    ulong contents_len = 0UL;
+
+    while( 1 ) {
+      if( FD_UNLIKELY( contents_len>=sizeof( contents ) ) ) FD_LOG_ERR(( "contents overflow" ));
+      long n = read( ctx->status_fds[ i ], contents + contents_len, sizeof( contents ) - contents_len );
+      if( FD_UNLIKELY( -1==n ) ) FD_LOG_ERR(( "read failed (%i-%s)", errno, strerror( errno ) ));
+      if( FD_UNLIKELY( 0==n ) ) break;
+      contents_len += (ulong)n;
+    }
+
+    int found_voluntary = 0;
+    int found_involuntary = 0;
+
+    char * line = contents;
+    while( 1 ) {
```

Comment:
I checked to see if there was a better way to get this info from the kernel, but it doesn't seem so.